Feature #3914 Set taxonomies from user defaults on login

dLobatog opened the pull request:
This PR allows users to have default taxonomies. These can be set on the users/edit form, although something like what @mccun934‎ suggested (http://projects.theforeman.org/attachments/download/654/org-selector-fav.png) would be nice to have in the future. I preferred to wait until inheritable taxonomies (https://github.com/theforeman/foreman/pull/1105) gets merged so that this can go in 1.4.

Also, I removed the restriction that only allowed admin users to clear their taxonomy selection. I think there's no reason to only let admins do that, not to mention that regular users could do this before this PR through the URL directly, this just exposes the option in the UI.

http://projects.theforeman.org/issues/3914

dLobatog committed:
Feature #3914 - Set taxonomies from user defaults

Migration for taxonomy foreign keys in User model

Default taxonomies relations in models

Non-admin users should be able to clear taxonomies

Users form to set default taxonomies

Set current taxonomies from user defaults on login

String extraction and main_app

Actually test default taxonomies are set on login

Taxonomies defaults are editable by current user

i18n labels

Merge error fix in User

Users API returns default taxonomies

Fix: Users can edit their own default taxonomies

Edit default taxonomies from select field

Default taxonomies in my account & hidden fields for admin user

Display full-title of nested organization

Styling issues & api docs

Use 'my_taxonomies' instead of custom method

Validators for User default_taxonomy

set_taxonomies on SSO authentication

Do not use camelCase for function names in js

Fixed validator

Test validator and style fixes

Admin default taxonomies can be anything

domcleal commented on app/views/home/_location_dropdown.html.erb:
```diff
@@ -1,17 +1,14 @@
+    <li><%= link_to(_('Any Location'), clear_locations_path) %></li>
+    <%= content_tag(:li, "", :class=>"divider") %>
     <% Location.my_locations.each do |location| %>
-      <li><%= link_to(location.name, main_app.select_location_path(location)) %></li>
```
main_app is required here when browsing a page from an isolated engine

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -47,6 +47,7 @@
             <%= f.select :locations_andor, [[_("must be"), "and"], [_("plus all"), "or"]], {}, :class => "form-control" %>
           </div>
           <%= multiple_checkboxes f, :locations, @user, Location, :label=>'' %>
+          <%= select_f f, :default_location_id, @user.locations, :id, :name, {:include_blank => true}, {:label => "Default Location"} %>
```
Use string extraction, e.g. `_("Default location")` and lower case the second word.

domcleal commented on test/functional/users_controller_test.rb:
```diff
@@ -245,4 +249,16 @@ def test_one #"should not remove the anonymous role" do
+      User.expects(:try_to_login).with(users(:one).login, users(:one).password).
+        returns(users(:one).post_successful_login)
+
+      post :login, { :login => { :login => users(:one).login, :password => users(:one).password } }
```
Should this verify the taxonomies are selected?

dLobatog replied:
Right, checking the defaults are set in the session is the most I can get without following the redirect, I can't directly check Taxonomy.current

domcleal replied:
That's good enough.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -47,6 +47,7 @@
             <%= f.select :locations_andor, [[_("must be"), "and"], [_("plus all"), "or"]], {}, :class => "form-control" %>
           </div>
           <%= multiple_checkboxes f, :locations, @user, Location, :label=>'' %>
+          <%= select_f f, :default_location_id, @user.locations, :id, :name, { :include_blank => true }, { :label => _('Default Location') } %>
```
Hmm, this is in the `unless @editing_self` section, which as an admin I'm not able to select my own default org/loc.

EDIT: or even a non-admin, it should be a user-level preference.

domcleal replied:
ping on this point?

dLobatog replied:
Latest update does not put this `select_f` in the `@editing_self` section so it should be available for everybody.

domcleal replied:
It doesn't appear to be, I guess because non-admins can't edit locations on themselves either, so the whole tab isn't shown.

dLobatog replied:
Sorry I just re-checked and it does work for non-admins, maybe you are using an user that doesn't have any permissions on locations/organizations? Try with an user with permissions view_locations and view_organizations.

domcleal replied:
That's a weird restriction then, perhaps we should automatically allow users to at least view locs/orgs that they're a member of?

dLobatog replied:
You mean giving users this role by default, or making it somehow unremovable?

domcleal replied:
Giving it by default seems a good idea, e.g. add the permission into the anonymous role?  As long as that means they can't see organisations they're _not_ a part of.

dLobatog replied:
Yeah, this role `view_organizations` just means the users can see organizations that have been previously assigned to them. I'll add it as a default so users can see this tab unless the role is removed.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -56,6 +57,7 @@
             <%= f.select :organizations_andor, [[_("must be"), "and"], [_("plus all"), "or"]], {}, :class => "form-control" %>
           </div>
           <%= multiple_checkboxes f, :organizations, @user, Organization, :label=>'' %>
+          <%= select_f f, :default_organization_id, @user.organizations, :id, :name, {:include_blank => true}, {:label => "Default Organization"}  %>
```
The UI's a little clunky here, as the dropdown menu isn't refreshed when you select organisations from the multiple_checkboxes above.  Can it refresh as that changes, or redesign it a little?

domcleal replied:
Missing `_()` around default org label.

dLobatog replied:
Mind if I create a redmine ticket for the ajax refresh thing and do that in a separate PR? 
I can do it too and I think it's a good idea, just wanted to focus on finishing http://projects.theforeman.org/issues/3214 for 1.4 in addition to this.

domcleal replied:
We can, though this PR won't go into 1.4 anyway, it's a feature & too late, so we can continue working on it if you're happy to.

dLobatog commented:
I've added `view_organizations` and `view_locations` as new permissions for the Anonymous role. This means users will be able to see the Locations/Organizations tab in their accounts at all times to be able to select a default only from the Locations/Organizations that have been assigned to them.

ehelms commented:
I thought I'd give this a whirl since it's pertinent to our usage of organizations in Katello. The tab allowing setting of a default (I wonder if we shouldn't just call it 'favorite') organization is inline with one way that we allowed users to set this in previous Katello. I am seeing the following:

1) As the admin user, and while editing the admin user while attempting to set a default organization I click submit and see a spinner and the no indication from the server that anything happened (but I see a 200 in my server logs). When I refresh the page I do not see my selection selected and if I log out and log back in my user does not automatically log into that organization.

2) During user creation, I see a dropdown to set my default organization but no choices in the dropdown. After thinking I have to select some organizations first, and selecting an organization the dropdown is still empty. Same is true if I edit my user after adding an organization to the user, the dropdown is not updated until I submit and refresh the page.

3) After logging in as a new user with a default organization set, I navigated to the my account page. The default organization dropdown presented me with a blank option and my current default organization. Changing the dropdown to the empty option, and clicking submit I was redirected to the 'hosts' page. After navigating back to my account, the previously set default organization was still chosen. This seems to be the case no matter if I choose a blank option or another organization.

4) I don't see the default organization listed as an attribute when asking the API for my user.

dLobatog commented:
@ehelms I think I've addressed all the points above, thanks for the feedback! Feel free to retest.

ehelms commented:
Looking good with respect to selecting and logging in under the selected default for admin and other users -- a few more things I noticed:

1) I would expect that setting the default organization for a user with multiple organizations should exist on the my account page. I created a new user with no permissions, gave them access to two organizations and after navigating to 'my account' was not able to set my own default organization.

2) I still think it's a strange interaction to add organizations to a user through the multiselect and not see the list of available organizations to set the default update. It means I have to:

a) Open user panel
b) Add organization 'Test Org'
c) Click 'submit'
d) Re-open the user panel
e) Click on 'Organization' tab
f) Select my default organization
g) Click 'submit'

3) Not sure if this is related directly to this PR, but I noticed the following:

a) Create two organizations
b) Add two organizations to a new user
c) Select organization A from the context switcher
d) Navigate to users and select the new user to edit
f) See the two organizations in the right hand multi-select, with the current organization A greyed out
g) Click on organization B to remove it (it now appears on the left hand side)
h) Click 'submit'
i) Switch to 'Any Context'
j) Open users panel and click edit on the new user
k) Notice that the user no longer has any organizations

dLobatog commented:
Point 1 is fixed, thanks.

Point 2 should be fixed after the fixes I made before, it should automatically load default taxonomies without clicking on submit, can you check the users.js javascript is being loaded when you edit an user in your setup (it is loaded on my setup).

Additionally I've set the locations/organizations/roles fields as hidden for the admin user.

ehelms commented:
Everything looks and works the way I would expect - just a few more edge case oddities:

1) For nested organizations, the organization multiselect shows the inheritance (e.g. ACME_Corp/Beta_Corp) but the default organization dropdown shows all organizations as flat (e.g. ACME_Corp, Beta_Corp, Another Org). Should these be consistently displayed?

2) This may be unrelated to this PR and a general bug -- the organization multi-select shows the current organization context greyed out when editing a user which implies that I shouldn't be able to de-select it but I can click on and remove it nonetheless.

3) The user is getting removed if trying to set the default organization to the current organization.

a) Add user to organization A
b) Select organization A from context selector
c) Navigate to users page
d) Select the user we added to organization A
f) Select organiation A from default organization dropdown and click save
g) Notice that user disappears from the list and no longer belongs to organization A

dLobatog commented:
1) I think showing the whole nested label is better. Fixed.
2) I noticed that too.. it's certainly not part of my PR, `git blame` shows @isratrade introduced it.. what was it supposed to mean? 
3) Can't replicate this. I added the user to an organization, set that organization as the 'current' organization, changed the default to that organization, and the user remains in the organization.

isratrade commented:
@eLobato, @ehelms, I'm looking into the issue of being able to de-select users even though they are disabled.  I ran into this in the past, but I thought it was fixed.

isratrade commented:
@eLobato, @ehelms, see PR #1282 for fix to deselecting disabled items

ehelms commented:
@eLobato The PR looks good to me functionality wise - ACK for all my points

domcleal commented on app/controllers/users_controller.rb:
```diff
@@ -115,7 +115,20 @@ def login_user(user)
     session[:user]         = user.id
     uri                    = session[:original_uri]
     session[:original_uri] = nil
+    set_current_taxonomies(user)
```
I think something similar will be needed for SSO?

dLobatog replied:
Yeah, [`authenticate`](https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/authentication.rb#L8) is where I think this actually belongs.

domcleal replied:
Yeah, in addition to here.  I don't think authenticate is used for logins via the login form.

domcleal commented on app/models/taxonomies/location.rb:
```diff
@@ -5,6 +5,7 @@ class Location < Taxonomy
   has_many_hosts :dependent => :nullify
 
   has_many :location_parameters, :dependent => :destroy, :foreign_key => :reference_id, :class_name => "LocationParameter"
+  has_many :default_users, :class_name => 'User', :foreign_key => :default_location_id
```
Nitpick, you neatly aligned the columns in organization.rb, but not here :)

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,75 @@
+                     { :include_blank => true }, { :label => _('Default organization') } %>
+
+      <%= selectable_f(f, :locale, FastGettext.human_available_locales,
+                        { :include_blank => _('Browser locale') } , { :label => 'Language' } ) %>
```
_() is missing from the label

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,75 @@
+                               { :disabled => @user.used_location_ids,
+                                 :label    => '' } ,
+                               { :onchange => 'taxonomyAdded(this, "location")' }
+                              ) unless @editing_self %>
```
Not required?

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,75 @@
+                                   { :disabled => @user.used_organization_ids,
+                                     :label => '' },
+                                   { :onchange => 'taxonomyAdded(this, "organization")' }
+                                  ) unless @editing_self %>
```
ditto

domcleal commented on app/views/home/_location_dropdown.html.erb:
```diff
@@ -1,12 +1,9 @@
-  <%= location_dropdown location_count %>
+  <%= location_dropdown Location.my_locations.count %>
   <ul class="dropdown-menu">
-    <% if User.current.admin? %>
```
I'm not sure this is a good idea, since #2298 is outstanding.

domcleal replied:
ping

dLobatog replied:
I can revert this, but users will then have to click on each organization to see their hosts. #864 addresses the issue anyway.

domcleal commented on app/views/api/v2/users/main.json.rabl:
```diff
@@ -3,3 +3,12 @@ object @user
 attributes :firstname, :lastname, :mail, :admin, :auth_source_id, :auth_source_name, :last_login_on, :created_at, :updated_at
+
+if SETTINGS[:locations_enabled]
+  attributes :default_location
```
[done] Needs adding to the apidocs too.

domcleal commented on app/models/user.rb:
```diff
@@ -13,25 +13,29 @@ class User < ActiveRecord::Base
   after_commit :ensure_default_role
 
   belongs_to :auth_source
+  belongs_to :default_organization, :class_name => 'Organization'
```
I don't see any validation of these attributes, could a malicious user set them (via the API or form POST) to gain access to an org of which they're not a member?

dLobatog replied:
Yeah, that'd be quite easy, I've added a validator that checks for inclusion in `Taxonomy.my_taxonomies`

domcleal commented on app/models/user.rb:
```diff
@@ -294,6 +298,10 @@ def taxonomy_foreign_conditions
   end
 
+  def available(model)
+    admin? ? model.all : self.send(model.to_s.tableize)
```
We have `Organization.my_organizations` and equivalent for this, which is already used in places like the context dropdown menu.

domcleal commented on app/assets/javascripts/users.js:
```diff
@@ -0,0 +1,31 @@
+function taxonomyAdded(taxonomies, type)
```
Nitpick, I think we generally prefer underscore style names rather than camel case.

dLobatog replied:
I thought so too, but then I saw most .js methods using camelCase, maybe I can change them all in this PR?

domcleal replied:
Let's not...

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,75 @@
+<%= javascript "users" %>
```
Try using the edit from the two-pane view on the users list and this JavaScript file isn't included it seems (taxonomyAdded is not defined.)

dLobatog replied:
That's weird, I used to get that error but it went away, I think an `assets:precompile` fixed it.

domcleal replied:
This was in development, so assets shouldn't be precompiled with the rake task.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,75 @@
+          <%= location_selects(f, @user.used_or_selected_location_ids,
+                               { :disabled => @user.used_location_ids,
+                                 :label    => '' } ,
+                               { :onchange => 'taxonomyAdded(this, "location")' }
```
This doesn't seem to work if you use the select/deselect all buttons.

dLobatog commented:
[test] , I'm not sure what is this breaking in Katello.

domcleal commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -32,6 +32,8 @@ def show
           param :admin, :bool, :required => false, :desc => "Is an admin account?"
           param :password, String, :required => true
+          param :default_location, String if SETTINGS[:locations_enabled]
+          param :default_organization, String if SETTINGS[:locations_enabled]
```
organizations_enabled

domcleal commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -74,6 +74,7 @@ def sso_authentication
 
   def set_current_user(user)
     User.current = user
+    Foreman::Controller::UsersMixin.set_current_taxonomies(user)
```
How well does this work with SSO?  Does it force their taxonomy context on every request back to the default?

dLobatog replied:
Why would it be reset? I think it'd be just kept in the session until it expires.

domcleal replied:
Ah yes, since this method sets the session[:user] we shouldn't reprocess this method on subsequent requests.

domcleal commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -32,6 +32,8 @@ def show
           param :mail, String, :required => true
           param :admin, :bool, :required => false, :desc => "Is an admin account?"
           param :password, String, :required => true
+          param :default_location, String if SETTINGS[:locations_enabled]
```
default_location_id and Integer, no?

dLobatog replied:
According to [this](https://github.com/theforeman/foreman/pull/1191/files#diff-c3845c26d7bd1558c3b4a76d711221e4R30) default_location and default_organization should be OK.

domcleal replied:
Those tests are setting objects, not strings.  The rest of the API docs use _id and integers.

domcleal commented on db/migrate/20140123185537_add_default_organization_id_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddDefaultOrganizationIdToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :default_organization_id, :integer
```
Needs a FK

EDIT: or maybe not.. should we have a `dependent => :nullify`?

domcleal replied:
ping, what happens on org/loc destroy?

dLobatog replied:
I tested this while logged in as an user with a default Organization, since the taxonomy doesn't exist anymore, it just gets removed from the default Organization.

domcleal replied:
Ah you're right, it's nullified already, cool.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,73 @@
 
-      <%= select_f(f, :auth_source_id, AuthSource.all.delete_if { |a| a.to_label.nil? }, :id, :to_label, {:include_blank => true}, {:onchange => 'auth_source_selected()', :label => _("Authorized by")}) unless @editing_self
-      %>
+      <%= select_f f, :default_location_id, Location.my_locations, :id, :title,
```
These shouldn't be on the form if locations/orgs are disabled.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,73 @@
-      <div id="password" <%= display? !@user.manage_password? %>>
-        <%= password_f f, :password, :label => _("Password") %>
-        <%= password_f f, :password_confirmation, :label => _("Verify") %>
+      <%= select_f f, :default_organization_id, Organization.my_organizations, :id, :title,
```
Is `my_organizations` correct?  Shouldn't it be the organizations visible to the user we're editing, not the current user?

domcleal commented on app/models/user.rb:
```diff
@@ -61,6 +65,8 @@ class User < ActiveRecord::Base
   validates :password_hash, :presence => true, :if => Proc.new {|user| user.manage_password?}
   validates_confirmation_of :password,  :if => Proc.new {|user| user.manage_password?}, :unless => Proc.new {|user| user.password.empty?}
   validates :firstname, :lastname, :format => {:with => /\A[[:alnum:]\s'_\-\.()<>;=,]*\z/}, :length => {:maximum => 50}, :allow_nil => true
+  validates :default_organization, :inclusion => { :in => Organization.my_organizations }, :allow_blank => true
```
ditto what I said on the form, my_organizations etc uses the User.current, not the user we're editing.

domcleal commented on app/models/taxonomies/organization.rb:
```diff
@@ -3,14 +3,16 @@ class Organization < Taxonomy
-  has_many :organization_parameters, :dependent => :destroy, :foreign_key => :reference_id, :class_name => "OrganizationParameter"
+
+  has_many :organization_parameters, :class_name => 'OrganizationParameter', :foreign_key => :reference_id, :dependent => :destroy
+  has_many :default_users,           :class_name => 'User',                  :foreign_key => :default_organization_id
```
dependent nullify?

domcleal commented:
I'd guess the katello failures during initialisation with a blank DB are due to the validators defined in the user model, which depend on Organization.my_organizations etc.  Maybe changing this to a validator method would be better, as at the moment I think the list of permitted organizations would be decided at class load time (hence why you added `if user.nil?` into these scopes) and not actually validated for the user.

dLobatog commented:
@domcleal It's green now, will this go in RC?

mccun934 commented:
+1 to 1.5 RC as much as possible.  We want this for Katello 1.5 which is being based on Foreman 1.5 and this feature is highly desirable to us.

domcleal commented:
Depends if I can review it in time, it's back on the list.

domcleal commented on app/models/taxonomies/organization.rb:
```diff
@@ -3,14 +3,16 @@ class Organization < Taxonomy
   scope :my_organizations, lambda {
       user = User.current
-      if user.admin?
+      if user.nil? || user.admin?
```
This can be reverted now the validation's fixed.  Ditto in location.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,77 @@
-      <%= select_f(f, :auth_source_id, AuthSource.all.delete_if { |a| a.to_label.nil? }, :id, :to_label, {:include_blank => true}, {:onchange => 'auth_source_selected()', :label => _("Authorized by")}) unless @editing_self
-      %>
+      <% if show_location_tab? %>
+      <%= select_f f, :default_location_id, @user.locations, :id, :title,
```
indentation's off

domcleal commented on app/models/user.rb:
```diff
@@ -403,4 +419,17 @@ def ensure_default_role
     self.roles << role unless self.role_ids.include?(role.id)
   end
+
+  def default_location_inclusion
```
Could you add unit tests?  I don't see any coverage.

dLobatog replied:
Done

domcleal commented on app/models/user.rb:
```diff
@@ -403,4 +419,17 @@ def ensure_default_role
   end
+
+  def default_location_inclusion
+    unless locations.include?(default_location) || default_location.blank?
```
I'm not sure `self.locations` will work properly with inheritance (ditto in the UI), I think you need to both list and permit children too.

dLobatog replied:
I agree, however the inheritance of locations/organizations and users seems a bit broken now. 

I have an user to which I have assigned location 'a'. I create several inherited locations whose parent is 'a', but the user locations don't get updated and hence the user cannot change its location to anything but 'a'. 

I'm all in for changing this and making it more consistent (I think the user should be able to access locations that inherit from 'a'). We used to handle a similar case for hostgroups (https://github.com/theforeman/foreman/blob/develop/app/controllers/hostgroups_controller.rb#L47) , in fact I think I found a regression bug there, we can probably update user.taxonomies in a similar fashion.

I added a test (commented out for the moment) to make sure we cover this at some point.

domcleal replied:
Hm ok, can you file bugs please to outline what's broken?

domcleal commented on app/assets/javascripts/users.js:
```diff
@@ -0,0 +1,31 @@
+function taxonomy_added(taxonomies, type)
```
Please add this file to the javascript list in config/environments/production.rb.

dLobatog replied:
Done

domcleal commented on app/models/user.rb:
```diff
@@ -403,4 +419,17 @@ def ensure_default_role
+  end
+
+  def default_organization_inclusion
+    unless organizations.include?(default_organization) || default_organization.blank?
```
`|| self.admin?` too, as admins don't get assigned to taxonomies, they have access to everything.

domcleal commented on app/views/users/_form.html.erb:
```diff
@@ -1,54 +1,77 @@
 
-      <%= select_f(f, :auth_source_id, AuthSource.all.delete_if { |a| a.to_label.nil? }, :id, :to_label, {:include_blank => true}, {:onchange => 'auth_source_selected()', :label => _("Authorized by")}) unless @editing_self
-      %>
+      <% if show_location_tab? %>
```
I think just `SETTINGS[:locations_enabled]` which I thought was in an earlier version, users can't change their default org at the moment without additional permissions.

domcleal commented:
:+1: tests well subject to the admin + form issues noted above.

dLobatog commented:
There you go.

domcleal commented:
Thanks @eLobato, merged as 7e1f0c7 for Foreman 1.5.0!

mccun934 commented:
:clap:
